src/Aspire.Dashboard/Otlp/Storage/ApplicationKey.cs

```diff
@@ -23,23 +23,30 @@ public bool EqualsCompositeName(string name)
             return false;
         }
 
-        // Composite name has the format "{Name}-{InstanceId}".
-        if (name.Length != Name.Length + InstanceId.Length + 1)
+        if (InstanceId != null)
         {
-            return false;
-        }
+            // Composite name has the format "{Name}-{InstanceId}".
+            if (name.Length != Name.Length + InstanceId.Length + 1)
+            {
+                return false;
+            }
 
-        if (!name.AsSpan(0, Name.Length).Equals(Name, StringComparisons.ResourceName))
-        {
-            return false;
+            if (!name.AsSpan(0, Name.Length).Equals(Name, StringComparisons.ResourceName))
+            {
+                return false;
+            }
+            if (name[Name.Length] != '-')
+            {
+                return false;
+            }
+            if (!name.AsSpan(Name.Length + 1, InstanceId.Length).Equals(InstanceId, StringComparisons.ResourceName))
+            {
+                return false;
+            }
         }
-        if (name[Name.Length] != '-')
+        else
         {
-            return false;
-        }
-        if (!name.AsSpan(Name.Length + 1, InstanceId.Length).Equals(InstanceId, StringComparisons.ResourceName))
-        {
-            return false;
+            return string.Equals(Name, name, StringComparisons.ResourceName);
```

Comment:
Does this need to consider when `other.InstanceId` is non-null?

Reply:
No. It's supposed to match other apps with the same name and multiple instances.

Reply:
Added a comment.